products/llbuildSwift/BuildDBBindings.swift

```diff
@@ -0,0 +1,345 @@
+#elseif os(Windows)
+import MSVCRT
+import WinSDK
+#else
+import Glibc
+#endif
+
+import Foundation
+
+// We don't need this import if we're building
+// this file as part of the llbuild framework.
+#if !LLBUILD_FRAMEWORK
+import llbuild
+#endif
+
+public typealias KeyID = UInt64
+public typealias KeyType = String
+
+public protocol BuildDBDelegate {
+    func getKeyID(key: KeyType) -> KeyID
+    func getKey(id: KeyID) -> KeyType
+}
+
+/// Defines the result of a call to fetch all keys from the database
+/// Wraps calls to the llbuild database, but all results are fetched and available with this result
+public class BuildDBFetchKeysResult {
+    /// Opaque pointer to the actual result object
+    private let result: OpaquePointer?
+    
+    fileprivate init() {
+        self.result = nil
+    }
+    
+    fileprivate init(result: OpaquePointer) {
+        self.result = result
+    }
+    
+    private lazy var _count: Int = result.map { Int(llb_database_result_keys_get_count($0)) } ?? 0
+    
+    /// Get the key identifier for a given key (is unique)
+    public func getKeyID(key: KeyType) -> KeyID {
+        assert(result != nil, "Can't get key identifier without fetching the data first.")
+        return llb_database_result_keys_get_id_for_key(result, key.cString(using: .utf8))
+    }
+    
+    /// Get the key for a given key identifier (might be unique)
+    public func getKey(id: KeyID) -> KeyType {
+        assert(result != nil, "Can't get key without fetching the data first.")
+        
+        var data = llb_data_t()
+        withUnsafeMutablePointer(to: &data) { ptr in
+            llb_database_result_keys_get_key_for_id(self.result, id, ptr)
+        }
+        defer { data.data?.deallocate() }
+        return stringFromData(data)
+    }
+    
+    deinit {
+        llb_database_destroy_result_keys(result)
+    }
+}
+
+extension BuildDBFetchKeysResult: Collection {
+    public typealias Index = KeyID
+    public typealias Element = (keyID: KeyID, key: KeyType)
+
+    public var startIndex: Index {
+        return 1 as UInt64
+    }
+    
+    public var endIndex: Index {
+        return UInt64(self.count) + startIndex
+    }
+    
+    public subscript(index: Index) -> Iterator.Element {
+        guard (startIndex..<endIndex).contains(index) else {
+            fatalError("Index \(index) is out of bounds (\(startIndex)..<\(endIndex))")
+        }
+        return (index, getKey(id: index))
+    }
+    
+    public func index(after i: KeyID) -> KeyID {
+        return i + 1
+    }
+}
+
+extension BuildDBFetchKeysResult: BuildDBDelegate {}
+
+/// Defines the result of a built task
+public class RuleResult {
+    /// Reference to the C API result instance (for desctruction)
+    private var result: llb_database_result_t
+    /// The value of the result
+    public let value: Data
+    /// Signature of the node that generated the result
+    public let signature: UInt64
+    /// The build iteration this result was computed at
```

Comment:
We use `timestamp` in other places versus iteration. I'd probably stay consistent, although one could argue we should change them all to a slightly better word like `epoch`.